src/interpreter/mod.rs

```diff
@@ -0,0 +1,492 @@
+              i', a', b', n' <== i + 1, b, c, n;
+             }
+            }
+
+            state middle {
+             signal c;
+
+             c <== a + b;
+
+             if i + 1 == n {
+              -> final {
+               i', b', n' <== i + 1, c, n;
+              }
+             } else {
+              -> middle {
+               i', a', b', n' <== i + 1, b, c, n;
+              }
+             }
+            }
+
+            // There is always a state called final.
+            // Output signals get automatically bindinded to the signals
+            // with the same name in the final step (last instance).
+            // This state can be implicit if there are no constraints in it.
+           }
+        ";
+
+        let compiled =
+            compile::<Fr>(code, Config::default(), &DebugSymRefFactory::new("", code)).unwrap();
+
+        let result = compiled
+            .wit_gen
+            .generate(HashMap::from([("n".to_string(), Fr::from(12))]));
+
+        println!("{:#?}", result);
+    }
+
+    #[test]
+    fn test_run_halo2_mock_prover() {
+        let code = "
+        machine fibo(signal n) (signal b: field) {
+            // n and be are created automatically as shared
+            // signals
+            signal a: field, i;
+
+            // there is always a state called initial
+            // input signals get binded to the signal
+            // in the initial state (first instance)
+            state initial {
+             signal c;
+
+             i, a, b, c <== 1, 1, 1, 2;
+
+             -> middle {
+              i', a', b', n' <== i + 1, b, c, n;
+             }
+            }
+
+            state middle {
+             signal c;
+
+             c <== a + b;
+
+             if i + 1 == n {
+              -> final {
+               i', b', n' <== i + 1, c, n;
+              }
+             } else {
+              -> middle {
+               i', a', b', n' <== i + 1, b, c, n;
+              }
+             }
+            }
+
+            // There is always a state called final.
+            // Output signals get automatically bindinded to the signals
+            // with the same name in the final step (last instance).
+            // This state can be implicit if there are no constraints in it.
+           }
+        ";
+
+        let mut chiquito =
+            compile::<Fr>(code, Config::default(), &DebugSymRefFactory::new("", code)).unwrap();
+
+        chiquito.circuit.num_steps = 12;
+
+        let witness = chiquito
+            .wit_gen
+            .generate(HashMap::from([("n".to_string(), Fr::from(12))]))
+            .unwrap();
+
+        let plonkish = chiquito.plonkish(config(
+            SingleRowCellManager {},
+            SimpleStepSelectorBuilder {},
+        ));
+
+        let compiled = chiquito2Halo2(plonkish.0);
+
+        let circuit = ChiquitoHalo2Circuit::new(
+            compiled,
+            plonkish.1.map(|g| g.generate_with_witness(witness)),
+        );
```

Comment:
Why do we need an intermediate `let compiled = chiquito2Halo2(plonkish.0);` instead of going from `plonkish.0` to `circuit` directly?

Reply:
Yes, this is design decision from the fact that you need the witness to create a halo2 circuit, when we compile to halo2 midleware perhaps we can make this interface more natural